torch/lib/c10d/logger.cpp

```diff
@@ -85,6 +87,101 @@ void Logger::set_construction_logging_data(
   LogPyTorchDDPUsage(*ddp_logging_data_);
 }
 
+void Logger::calculate_avg_cpu_time(
+    int64_t& avg_time,
+    int64_t cpu_start_time,
+    int64_t cpu_end_time) {
+  long num_iters = reducer_->num_iterations_;
+  avg_time = ((cpu_end_time - cpu_start_time) + avg_time * (num_iters - 1)) /
+      num_iters;
+}
+
+#ifdef USE_CUDA
+void Logger::calculate_avg_gpu_time(
+    int64_t& avg_time,
+    cudaEvent_t gpu_start,
+    cudaEvent_t gpu_end) {
+  long num_iters = reducer_->num_iterations_;
+  float milliseconds = 0.0;
+  cudaEventElapsedTime(&milliseconds, gpu_start, gpu_end);
+  avg_time =
+      (int(milliseconds * 1000000) + avg_time * (num_iters - 1)) / num_iters;
+}
+#endif
+
+void Logger::set_runtime_stats() {
+  // set runtime stats after 1st iteration is complete.
+  if (reducer_->num_iterations_ == 0) {
+    return;
+  }
+  // Set ith iteration when the runtime stats are set.
+  ddp_logging_data_->iteration = reducer_->num_iterations_;
+  // If unused_parameters_ is not empty, calculate its sizes.
+  // unused_parameters_ is calculated in forward call of
+  // each iteration.
+  for (const auto& unused_index : reducer_->unused_parameters_) {
+    const auto& v = reducer_->replicas_[unused_index.replica_index]
+                                       [unused_index.variable_index];
+    ddp_logging_data_->unused_parameter_size += v.numel() * v.element_size();
+  }
+  // rebuilt_bucket_sizes will not change once buckets are rebuilt,
+  // so it only needs to set once during whole training loop.
+  if (reducer_->num_iterations_ == 2) {
```

Comment:
This seems like it could be bugprone since we are 1) hardcoding the iteration number and 2) Rebuild bucket logic can change in DDP, and this could start to silently be incorrect. 

Could we just keep `ddp_logging_data_->has_rebuilt_buckets ` in sync with `reducer_->has_rebuilt_bucket_;` and set the bucket sizes whenever the latter changes?

Reply:
good idea!